Create parser module and create schedule parser

AllanWang opened the pull request:
Resolves #91 

Changes are generally detailed within the commit messages.
Pending PR so you can review as you go.

jguerinet commented:
This is a good start. I'd rather have a bunch of small PRs that we merge as we go instead of 1 big PR with a whole bunch of changes. Does that work for you?

AllanWang commented:
I'm good with that. I'm going to move the new parser changes to another module since we agree on that. I'll add tests from redacted html source just to make sure that nothing breaks.

Can you elaborate on what converters broke from jsoup 1.10.x? Ideally I'd write the new module with the newest jsoup

AllanWang commented:
This PR is now significantly bigger, but it's because of the new `parser` module and the new tests.

The code from the first change has been commented out and moved to the new module, so this PR does not touch the `app` module at all.

The PR deals with parsing the student schedule.

Parser changes:
* Initial selector tag changed from `datadisplaytable` to `table.datadisplaytable`
* Parser is written in a way where it should never crash. Any number parses either fall back to `null` and return prematurely on error, or are matched against a regex.
* All new parse methods take in a `ParseDebugger`. The debugger allows the parser to produce debug messages, or messages explaining why the parser did not match. This is useful in figuring out which part fails when the minerva layout changes. By default, the messages are ignored. In the test code, they are added to a list and later checked to be empty. In android, you can use it to send events to Fabric.

To get parsers to work, relevant models need to be copied over. Eventually, all parse models will end up in this module and imported to `app`.

Changes:
* `Term` is now comparable, instead of having an `isAfter` function.
* Term is now a data class, and has data equality be default
* `Season` values are now reordered so that it's comparable feature is as expected. It used to be fall, winter, summer, and is now winter, summer, fall
* `Season` now has `number` set per enum, rather than a `number` attribute using a switch statement. This isn't related to this module, but not that seasons should be compared directly now, rather than converting their associated numbers to ints and comparing those values

---

Design wise, you'll notice that all parsers are global functions, and are internal. They are only exposed through the `MyMartletParser` object, to avoid cluttering projects that use it. The functions are internal so they can still be used within the module.

---

Changes that need to be addressed to use `parser` in `app`.

* Some models contain functions that require `Context` to get string resources, or  have associated resource ids. To make the models purely in Kotlin, those parts are removed. They can be added through extension functions within android modules in the future.
* `Course` is now a data class, whereas it used to be an `open class`. Unless there's a reason `CourseResult` has to extend `Course`, they can be independent. They are parsed completely differently anyways.
* All models in `app` are annotated with dbflow related annotations. As the core dependency doesn't depend on android, you can still add that in `parser`. However, it is sort of irrelevant to that module.
* Eventually, all of retrofit should probably be moved into this module as well. That way you'd be able to handle all the converters and run tests on new minerva pages from this module.

jguerinet commented:
A few comments: 
* CRN is indeed unique only within a semester, so concatenating it with the term _should_ work (we'll have to double check this)
* `Course` was open for `CourseResult` to cut down on duplicate properties, but they are significantly different that they can be separate objects. 
* Let's put in comments within the test classes as well. Simple ones, that just explain what you're actually testing. Class comments should also have the `@author` and `@since` annotations so we know when we started testing these things, and who wrote it.

jguerinet commented:
Furthermore: 
* Point your branch to `develop` instead of `master`. That should fix the Travis problem (all future branches should be to `develop`. 
* I don't remember exactly why JSoup was failing. If you switch to the latest version and run it you'll very quickly see the problem. If I recall, some of the tables were not being read correctly, or being read empty? I can't remember for sure.

AllanWang commented:
@jguerinet I've addressed the issues by adding docs to all non test methods. It isn't very useful to say that the code is 'since 2.3.2' though when it isn't actually added.

As you see above, you may also want to consider disabling the bot comments, or somehow limiting the number of comments it can have per PR. Some of the issues are important, but a lot of them are also irrelevant, especially for the test package.

jguerinet commented:
Yeah, I turned that feature on to see happened and I don't like it one bit.. I'll turn it off again later today.

On Jan. 15, 2019 06:54, Allan Wang <notifications@github.com> wrote:

@jguerinet<https://github.com/jguerinet> I've addressed the issues by adding docs to all non test methods. It isn't very useful to say that the code is 'since 2.3.2' though when it isn't actually added.

As you see above, you may also want to consider disabling the bot comments, or somehow limiting the number of comments it can have per PR. Some of the issues are important, but a lot of them are also irrelevant, especially for the test package.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/jguerinet/MyMartlet/pull/94#issuecomment-454275628>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AFLAKQpI4GrflJBnu68STv8RKIreckDZks5vDW0BgaJpZM4Z2p3s>.

jguerinet commented on parser/src/main/kotlin/com/guerinet/mymartlet/model/Course.kt:
```diff
@@ -0,0 +1,75 @@
+
+package com.guerinet.mymartlet.model
+
+import java.time.DayOfWeek
```
You're going to need to use the Backport here, as the Java 8 Date API isn't available on every Android version (yet).

AllanWang replied:
I'll add it later today. As a note, [threetenbp](https://github.com/ThreeTen/threetenbp) is added through [ThreeTenABP](https://github.com/JakeWharton/ThreeTenABP), whereas there will be no android dependencies in this module. I will add the latest version, but once parser gets used in app, the version may need to be downgraded to match. You can probably ignore the package from the import if the API hasn't changed.

jguerinet commented on parser/src/main/kotlin/com/guerinet/mymartlet/parser/Schedule.kt:
```diff
@@ -0,0 +1,162 @@
+    val row = getOrNull(index) ?: return notFound("Elements is empty")
+    val nextRow = getOrNull(index + 1) ?: return notFound("Element only has one row")
+    // Fast track; not expected data
+    if (!row.`is`(SCHEDULE_TABLE_QUERY) || !nextRow.`is`(SCHEDULE_TABLE_QUERY))
```
Curly braces for every `if` statement, even if only 1 line long.

jguerinet commented on parser/src/test/kotlin/com/guerinet/mymartlet/model/SeasonTest.kt:
```diff
@@ -0,0 +1,37 @@
+
+/**
+ * Tests for [Season]
+ */
```
Throw in the author tag here.

AllanWang commented:
Requested changes have been added. Yes, this is currently not integrated with the module. I'd recommend the entire parser be ported over and tested before we start mapping data models back to the app and adding the additional methods. The first stage now is separating the parsing logic from the app logic.

jguerinet commented:
Is this good to be re-reviewed? If so in the future can you assign the PR back to me so I know?

AllanWang commented:
Yeah it's good. Not used to github's review process. If you want, in the future you can also ask for my reviews if you want someone else to look through your changes. I can probably review at least once a day

jguerinet commented:
Sounds good, I'll start assigning some stuff to you then, thanks!
